### PR TITLE
Backend promo banner alterations

### PR DESCRIPTION
**Pull request by DamienBurkeWN**
### Setup
Check out this branch. Do a **drush cim** to import the config changes (and drush updb?).

### Testing
1. In the CMS, edit the homepage and go to the News & Promotions tab. Add as many promotion banners as you are allowed - should stop at 4.
2. Go to https://core-cms.northants.localhost/graphql and enter this query: 
 `{
  route(path: "/") {
    node {
      title
      promoBanner {
        id
        title
        body {
          value
        }
        link {
          url
          title
        }
        imageLarge
        imageMedium
        imageSmall
        imageAltText
      }
    }
  }
}`
Expect to see data matching what you entered.

Note that the frontend will not work on the home page due to these changes - further PR for frontend changes necessary: https://github.com/FutureNorthants/northants-website/pull/191

### CODE REVIEW

File: backend/src/config/sync/core.entity_form_display.node.homepage.default.yml [hunk 1]
```diff
@@ -88,18 +88,19 @@ third_party_settings:
       label: Hero
     group_promotional_banner:
       children:
+        - field_number_of_news_stories
```
Comment (DamienBurkeWN) [inline backend/src/config/sync/core.entity_form_display.node.homepage.default.yml:91]: Just moved this to the promo tab...

File: backend/src/config/sync/core.entity_form_display.node.homepage.default.yml [hunk 2]
```diff
@@ -88,18 +88,19 @@ third_party_settings:
...
       format_settings:
-        id: ''
-        classes: ''
         description: ''
         formatter: closed
         required_fields: true
-      label: 'Promotional Banner'
+        id: ''
+        classes: ''
+      label: 'News & Promotions'
```
Comment (DamienBurkeWN) [inline backend/src/config/sync/core.entity_form_display.node.homepage.default.yml:103]: ...and renamed the tab

File: backend/src/config/sync/field.storage.node.field_promotional_banner.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ settings:
   target_type: paragraph
 module: entity_reference_revisions
 locked: false
-cardinality: 1
+cardinality: 4
```
Comment (DamienBurkeWN) [inline backend/src/config/sync/field.storage.node.field_promotional_banner.yml:17]: Allows 4 promo banners to be added

File: backend/src/modules/nc_system/src/Entity/Node/Homepage.php [hunk 4]
```diff
@@ -144,16 +143,19 @@ public function getMemorialIconValue()
     return !empty($field) && isset($field[0]['value']) ? $field[0]['value'] : null;
   }
 
-  public function getPromoBanner(): ?PromoBanner
+  /**
+   * Returns the promotion paragraphs (up to 4) that make up the promotional banner area of the home page
+   */
+  public function getPromoBanner(): array
```
Comment (DamienBurkeWN) [inline backend/src/modules/nc_system/src/Entity/Node/Homepage.php:149]: Now returning the whole array of promo banners, not just the first instance

File: backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php [hunk 5]
```diff
@@ -23,48 +23,89 @@ public function getBody(): TextItemBase {
     return $bodyItem;
   }
 
+  public function getResolvedBody() {
```
Comment (DamienBurkeWN) [inline backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php:26]: Mostly tidying and refactoring for future maintainability in here, as previously done on Node types

File: backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <?php
-
```
Comment (DamienBurkeWN) [inline backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php:2]: Rather messed up looking diff in github, best looking at the raw file too

File: backend/src/schema.graphql [hunk 7]
```diff
@@ -492,8 +492,10 @@ type PromoBannerParagraph implements Paragraph
   title: String!
   body: FormattedField!
   link: LinkFieldWithTitle!
-  image1440x810: Image!
-  image144x81: Image!
+  imageLarge: String!
```
Comment (DamienBurkeWN) [inline backend/src/schema.graphql:495]: Introducing more generic image field names, 3 sizes now in use for promos (banner large & lazy loading small, tile medium & lazy loading small)

File: backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php [hunk 8]
```diff
@@ -9,62 +9,100 @@
 
 class PromoBanner extends Paragraph implements GraphQLEntityFieldResolver {
 
-  public function getTitle(): string {
+  protected function getTitle(): string {
```
Comment (chrisrhymes) [inline backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php:12]: Shouldn't the opening `{` go on a new line? Seems to be inconsistent in this file?

**Comment by chrisrhymes** [comment]
Is it possible to add a character limit to the body for a promotional banner item? It could make the layout very ugly if the body is very long in a promotional banner.

**Comment by DamienBurkeWN** [comment]
> .

From a brief look there doesn't seem to be a way to do that in Drupal without writing some code. @WestNorthantsRob is this something where we can just rely upon the common sense of content editors?

File: backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php (hunk 8 above)
Comment (DamienBurkeWN) [reply backend/src/modules/nc_system/src/Entity/Paragraph/PromoBanner.php:12]: Drupal standard is on the same line, I've fixed the ones where it was a on a new line instead

**Comment by WestNorthantsRob** [comment]
Hi Damien

You’re asking a lot but it makes sense.

**Review by chrisrhymes — APPROVED** [review]

**Pull request merged**